src/recommendationservice/logger.py

```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/python
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import sys
+from pythonjsonlogger import jsonlogger
+
+# TODO(yoshifumi) this class is duplicated since other Python services are
+# not sharing the modules for logging.
+class CustomJsonFormatter(jsonlogger.JsonFormatter):
```

Comment:
Is it possible to avoid this file altogether? (since we're still duplicaing it.)
I want to think initializing a SD-compatible logging class shouldn't be 40 lines.

If there's any examples of logging to SD from 12-factor Python apps, I think we should try to reuse those patterns.

Reply:
(Sorry I squashed the commits for better commit logs)

I agree that it would be nice if we can use a neat logging library which enable us to create SD compatible JSON structured logs. I'm still searching the nicer library than python-json-logger but it seems that's the best option so far.

One thing I want to emphasize here is that in Python it's really common to use `logging` standard library and create custom Handler or Formatter for desired log format, even if the format is flat text format, because in Python logging lib, the default format of log levels are in integers and level names are alias of those. In order to extract log level string, we need to have some custom class to extract `levelname` data from `record` object.

Also, the duplication happened due to the project structure. If we can share the module across all Python services, this problems never happens. Moreover, as you say, it's best for SD Logging to provide one-stop library for this use case.

Reply:
I see. I just started a thread internally to see maybe if there's an utility library we're missing. Truth to be told, this was much easier in Go. Initializing a JSON logger took 2 lines and 5 more lines to specify the custom fields. (We didn't even need to create a new method for it.) Let's wait and see what comes out of internal thread.